### PR TITLE
AK: find_last in String[View] using StringView

### PR DESCRIPTION
**Pull request by ldm5180**
Problem:
- `find_last` exists for finding single characters, but not for
  finding a string.

Solution:
- Add functionality to StringUtils for finding a StringView as a
  substring of a StringView.
- Plumb this through to provide String and StringView member
  functions.

### CODE REVIEW

File: AK/StringUtils.cpp [hunk 1]
```diff
@@ -351,6 +351,23 @@ Optional<size_t> find_last(StringView const& haystack, char needle)
...
+{
+    if (needle.length() > haystack.length()) {
+        return {};
+    } else if (needle == haystack) {
+        return 0;
+    }
+
+    for (auto i = haystack.length() - needle.length() + 1; i > 0; --i) {
+        auto const subhaystack = haystack.substring_view(i - 1, needle.length());
+        if (subhaystack.length() >= needle.length() && subhaystack == needle) {
```
Comment (mustafaquraish) [inline AK/StringUtils.cpp:364]: `subhaystack.length()` would never be greater than `needle.length()`, by definition, right? Perhaps just:

```cpp
if (subhaystack == needle) {
```
would be sufficient.

File: AK/StringUtils.cpp [hunk 2]
```diff
@@ -351,6 +351,23 @@ Optional<size_t> find_last(StringView const& haystack, char needle)
...
 
+Optional<size_t> find_last(StringView const& haystack, StringView const& needle)
+{
+    if (needle.length() > haystack.length()) {
+        return {};
+    } else if (needle == haystack) {
+        return 0;
+    }
+
+    for (auto i = haystack.length() - needle.length() + 1; i > 0; --i) {
```
Comment (mustafaquraish) [inline AK/StringUtils.cpp:362]: You start your loop from `i = haystack.length() - needle.length() + 1`, but then only ever use `i - 1` inside the loop. Is there a need to offset `i` by 1 that I'm not seeing?

File: AK/StringUtils.cpp (hunk 1 above)
Comment (ldm5180) [reply AK/StringUtils.cpp]: Thanks for reviewing and catching this!

File: AK/StringUtils.cpp (hunk 2 above)
Comment (ldm5180) [reply AK/StringUtils.cpp:362]: I think this is required unless you notice a different way. The `- 1` is needed `i` is unsigned. The loop will stop when the index is `0` and never be able to match a string with starts at the beginning.

Because the `- 1`, the `+ 1` is needed so matches can be made at the end of the string.

These boundary conditions are tested in the unit tests. If you can find a better way, I am open to it.

Comment (mustafaquraish) [reply AK/StringUtils.cpp:362]: Ah I see, makes sense. I would have just done

```cpp
for (int i = haystack.length() - needle.length(); i >= 0; --i) {
    // Use `i`
```

But I guess a narrowing conversation to `int` isn't ideal now that you point it out

**Comment by bgianfo** [comment]
Do we have an actual consumer of this API? As you know we don't normally merge APIs additions without callers.

**Comment by ldm5180** [comment]
I don’t think so. There was a FIXME to add it.
